### PR TITLE
Added support for brotli ('br') content-encoding

### PR DESCRIPTION
**Pull request by danielgindi**
https://github.com/expressjs/body-parser/pull/403

https://medium.com/oyotech/how-brotli-compression-gave-us-37-latency-improvement-14d41e50fee4
https://caniuse.com/#feat=brotli

### CODE REVIEW

File: test/compression.js
Comment (dougwilson) [inline test/compression.js:676]: Instead of excluding the test, have the test always be there but show up as skipped in the mocha output when the support is missing.

File: test/compression.js
Comment (dougwilson) [inline test/compression.js:522]: Instead of excluding the test, have the test always be there but show up as skipped in the mocha output when the support is missing.

File: test/compression.js
Comment (dougwilson) [inline test/compression.js:478]: Instead of excluding the test, have the test always be there but show up as skipped in the mocha output when the support is missing.

File: index.js
Comment (dougwilson) [inline index.js:88]: This needs to be configurable by the user.

File: index.js
Comment (dougwilson) [inline index.js:64]: Pull in a module to do this if it is necessary; let's not copy and paste a snippet that will be flagged as prototype pollution vulnerability.

File: test/compression.js
Comment (dougwilson) [inline test/compression.js:18]: The process module is just a global so no need to require it or have in a try catch.

File: index.js
Comment (dougwilson) [inline index.js:48]: The process module is just a global so no need to require it or have in a try catch.

File: index.js
Comment (danielgindi) [reply index.js:88]: > This needs to be configurable by the user.

It is. It's just a default (hinted by node.js docs but not set as the default. For some reason the highest level 11 is the default, which takes enormous amount of time with not a big size benefit.).
Level 4 is the sweet spot, kind of gzip's default just much better as it's brotli...

Comment (dougwilson) [reply index.js:88]: Apologies! Please add to the documentation how the user can configure this, then.

**Comment by danielgindi** [comment]
@dougwilson There's an issue with `object-assign` polyfill - it does not support the *very* outdated versions of node.js

**Comment by dougwilson** [comment]
> There's an issue with object-assign polyfill - it does not support the very outdated versions of node.js

I guess just pick a different module or an older version of that module.

File: index.js
Comment (dougwilson) [inline index.js:25]: Keep these in alpha-order

File: package.json
Comment (dougwilson) [inline package.json:16]: Dependencies outside of our org(s) need to have their version number pinned.

File: test/compression.js
Comment (dougwilson) [inline test/compression.js:16]: It looks like the code coverage only shows one of these branches are ever run. Under what Node.js versions would the other branch run?

File: index.js
Comment (dougwilson) [inline index.js:45]: It looks like the code coverage only shows one of these branches are ever run. Under what Node.js versions would the other branch run?

File: index.js
Comment (danielgindi) [reply index.js:73]: :-) Done!

File: test/compression.js
Comment (danielgindi) [reply test/compression.js]: Yeah there were errors in the tests before, with process module. But I guess it was because I `require`d it. Removing the `require` changed the coverage, so it must be it. I'm cleaning up a bit now.

File: README.md
Comment (dougwilson) [inline README.md:52]: Move this into the documentation section of the option that the user will use to change the value, rather than just at the top.

Comment (danielgindi) [reply README.md:52]: The documentation is in Node's docs. There's no point in copying those options here...

Comment (dougwilson) [reply README.md:52]: Can you at least show me an example for how to change the level through this module's interface? I cannot figure it out, at least.

Comment (dougwilson) [reply README.md:52]: For example, is it not the existing [level option](https://github.com/expressjs/compression#level) of this module?

File: README.md
Comment (dougwilson) [inline README.md:16]: Everything I am reading about br says that compression speeds are significantly slower than gzip. Can you link a benchmark (ideally using Node.js itself, since that is what we are using) that shows that br is faster than gzip and deflate? I'm only asked because I just know from history a claim like this is the readme is going to attract attention and so we could either have the benchmarks to show it or just not even make such a statement in the readme of this module.

Comment (danielgindi) [reply README.md:16]: I’ll try to add a link tomorrow or the day after, as we are moving to a different apartment tomorrow.
Anyway there are lots of resources, and I did a benchmark myself on friday morning using node.js. The results were outstanding for text and binary alike. Compressed size was 10% than the gzip (comparing level 4 against gzip’s best!), and speeds were between 14 to 27 times faster, depending on compression levels.

Actually Zstd (facebook) is even faster (especially in decompression) and have even better compression ratios, but it has no up to date native implementation for node, and of course no built in support.
The benchmarks in node with the existing zstd modules basically suck (speed-wise).

Some links:

https://cran.r-project.org/web/packages/brotli/vignettes/brotli-2015-09-22.pdf

http://www.instantshift.com/2018/03/02/gzip-vs-brotli-compression/

https://expeditedsecurity.com/blog/nginx-brotli/

https://medium.com/oyotech/how-brotli-compression-gave-us-37-latency-improvement-14d41e50fee4

Comment (danielgindi) [reply README.md]: I've updated the docs.

File: README.md
Comment (dougwilson) [inline README.md:16]: I don't think this makes the statement any more clearer. I would rather we just leave out editorial comments like this all together and just keep it to the facts of the module -- that br is only supported in specific modules. Even if the editorial was left in, what good does it serve the users of this module? There is no option for them to disable or enable br and the usage of it is left completely up to the clients accessing the server. Besides, this link demonstrates how to use br with a completely different module than this module, which is just going to be confusing.

File: README.md
Comment (dougwilson) [inline README.md:53]: Move this down into the documentation for setting the br level.

File: README.md
Comment (dougwilson) [inline README.md:55]: Please choose a different link for this, as this link demonstrates how to use br with a completely different module than this module, which is just going to be confusing.

File: README.md
Comment (dougwilson) [inline README.md:152]: These headings should be in alpha order

File: README.md
Comment (dougwilson) [inline README.md:154]: It's quite unclear from this documentation what `params` is supposed to take as a value... is this a list or something? Can you add a plain description around this option and what it does, etc. like the other options?

File: index.js
Comment (dougwilson) [reply index.js:73]: This is not resolved still. It seems like it's not possible to change the level still. If you disagree, please post an example program here that demonstrates this module with the level set to 1.

**Comment by danielgindi** [comment]
@dougwilson You just pass them as `compression({ params: { [zlib.constants.BROTLI_PARAM_QUALITY]: 8 } })`.

File: README.md
Comment (danielgindi) [reply README.md:154]: The `params` are extra parameters object. That's how the zlib module is defined in node.
Anyone playing with the parameters is actually using zlib options, and is either familiar with it or going to read the docs. There's no point in recreating node's docs in here.

File: index.js
Comment (danielgindi) [reply index.js:73]: You just pass them as `compression({ params: { [zlib.constants.BROTLI_PARAM_QUALITY]: 8 } })`.

Comment (danielgindi) [reply index.js:73]: This piece of code just adds a reasonable default. That's it. It only does so if the `params` option is `undefined`.

File: README.md
Comment (dougwilson) [reply README.md:154]: Yes, there is a point. The reason they are recreated is because I had the same attitude, but after the x-th issue opened in this repo asking the same question, it is worth recreating all the docs here. If you disagree, you're always welcome to host, maintain, and triage incoming issues on your own module :) Please write out all the docs for all options that this module is exposing directly in the README.

File: index.js
Comment (dougwilson) [reply index.js:73]: Wouldn't the user expect passing in `compression({ params: { foo: bar } })` still have a quality of 4 set? This does not seem to do that, and the documentation doesn't indicate that if you're passing any other params that the default quality is no longer 4. That is either a miss on the documentation or a miss on the code for that.

File: README.md
Comment (danielgindi) [reply README.md]: Cleared.

File: README.md
Comment (danielgindi) [reply README.md]: I don't have a different link and honestly don't have the patience to look for one. I did my own benchmarks, I've seen this one who also did benchmarks, and node's examples also use level 4, but the actual default there is an unusable 11. I'm stripping the link.

File: index.js
Comment (danielgindi) [reply index.js:73]: It's a miss in the code after refactoring out the original shallow copy code. Adding a fix.

**Comment by danielgindi** [comment]
> I added a comment about the faster statement and still have an open question on how a user can change compression level of br. I'm also having trouble to actually get br compression to even work at all with Chrome. I'm trying to figure it out, as our number 1 issue opened here is how to get this module working, so having concrete information for how to get br working with a web browser (Chrome, for instance) would help a lot. For reference I used the example in the README and your branch and Chrome continues to only show it using gzip, even when the connection is https (which my understanding is a requirement for br to work in Chrome).

Look at Chrome's `Accept-Encoding: gzip, deflate, br` - it puts `br` last. I'm guessing they put it last as a transition period, as [some poor servers out there fail when brotli is specified](https://support.google.com/websearch/thread/13460143?hl=en), or [intermediaries doing even worse](https://bugs.chromium.org/p/chromium/issues/detail?id=452335#c87).

Or maybe it's because when it first arrived they considered it a good compression for WOFF fonts, but they always tested with the highest compression levels. Today people know that with level 4 you have better results on all kinds of files.

**Comment by dougwilson** [comment]
So what clients will end up using br with this PR I can test?

**Comment by danielgindi** [comment]
> So what clients will end up using br with this PR I can test?

If you'll investigate the headers of css/js files in [this url](https://tools.keycdn.com/brotli-test), you'll see that they're encoded with brotli. They just prioritize brotli over gzip server-side. Which is [the right thing](https://engineering.linkedin.com/blog/2017/05/boosting-site-speed-using-brotli-compression) to do, like we already do for gzip over deflate.

**Comment by danielgindi** [comment]
What I mean to say is, that you can temporarily disable gzip in the code or prioritize br, to make chrome communicate with your server using brotli.

**Comment by Fonger** [comment]
@danielgindi I agree with you.

1. Cloudflare, the largest CDN in the world, prioritizes brotli over gzip at server-side.
2. [koa-compress](https://github.com/koajs/compress) has recently added brotli support and they prioritize brotli over gzip server-side too.
https://github.com/koajs/compress/blob/80054a551f898d371c60dcb17b749da61cf02b5a/lib/encodings.js#L27

**Comment by danielgindi** [comment]
How are we doing?

**Comment by dougwilson** [comment]
Waiting for your changes to review.

**Comment by danielgindi** [comment]
> Waiting for your changes to review.

Anything specific?

**Comment by NathanielHill** [comment]
any updates on when this will get merged and released?

**Comment by dougwilson** [comment]
Hi @NathanielHill I am just waiting on the changes referenced above, specifically, when I tested this PR, it did not result in br responses in any of the major web browsers (Chrome, Firefox) that support br, so I'm kind of at a loss for what value this change would bring or if it is broken.

If this change is not going to actually result in the (better) br compression in the web browsers, perhaps someone can better explain what the purpose of this PR is? FWIW, the Express middleware `shrinkray` does result in br in those web browsers, so it's certainly possible to achieve.

**Comment by devknoll** [comment]
I think this might need a test case for `when "Accept-Encoding: gzip, deflate, br"` -- this is what my Chrome is reporting, at least. Seems like maybe gzip is being preferred when both are provided due to the ordering in the arrays?

**Comment by danielgindi** [comment]
> Hi @NathanielHill I am just waiting on the changes referenced above, specifically, when I tested this PR, it did not result in br responses in any of the major web browsers (Chrome, Firefox) that support br, so I'm kind of at a loss for what value this change would bring or if it is broken.
> 
> 
> 
> If this change is not going to actually result in the (better) br compression in the web browsers, perhaps someone can better explain what the purpose of this PR is? FWIW, the Express middleware `shrinkray` does result in br in those web browsers, so it's certainly possible to achieve.

Of course it's possible to "achieve". But I did not want to presume that I can change the prioritization here. Gzip is currently prioritized in the code.
Currently browsers do not put `br` first, but there are other clients (i.e api requests between servers).

If this is what you want (which is not clear at all, as you just kept your silence about this), it's easy to achieve.

**Comment by dougwilson** [comment]
Hi @danielgindi sorry if I wasn't clear. I tried to start that conversation in https://github.com/expressjs/compression/pull/172#pullrequestreview-447507768 but you never seemed to answer my question. So perhaps we should circle back around to that, and can you provide what I asked for? :

> I'm also having trouble to actually get br compression to even work at all with Chrome. I'm trying to figure it out, as our number 1 issue opened here is how to get this module working, so having concrete information for how to get br working with a web browser (Chrome, for instance) would help a lot.

You never gave the name of any web browser where this actually works. The number 1 issue here is how to get this module working. If this doesn't actually work out of the box, then we need to understand why and determine what (if any) changes need to be made and what the concrete value it adds.

Perhaps we can also ask the folks that commented here: @Fonger would you expect this PR to get br in a web browser? @NathanielHill is that the same expectation you have as well and want to see released?

**Comment by NathanielHill** [comment]
@dougwilson Yes, I would expect a server to send `br` over `gzip` if both are present in `Accept-Encoding`

**Comment by Fonger** [comment]
@dougwilson Yes

If I see accept-encoding `gzip, defalte, br`, which is what modern browsers currently use, this middleware should use `br` instead of `gzip`.

**Comment by danielgindi** [comment]
> Hi @danielgindi sorry if I wasn't clear. I tried to start that conversation in https://github.com/expressjs/compression/pull/172#pullrequestreview-447507768 but you never seemed to answer my question. So perhaps we should circle back around to that, and can you provide what I asked for? :

Well I did answer that, but perhaps the answer is not so simple.
By standard you should respect the requested prioritization, and currently browsers prefer gzip for some reason. I guess there is an assumption that a lot of servers do not support br, so br was added without much thought.

Anyway this module already re-prioritizes gzip over deflate. So if that's what we're doing, we should also do this for br over the rest.

**Comment by danielgindi** [comment]
@dougwilson I've added a commit that re-prioritizes the encodings to br-gzip-deflate.
Will work with any browser now.

**Comment by nicksrandall** [comment]
@dougwilson are you available to re-review this PR? I'd love to see this get merged sooner rather than later. Thanks for all you do!

**Comment by dougwilson** [comment]
Hi @nicksrandall yes, I am. It is on my to-do list, but between @danielgindi 's note and now, there have been 3 security vulnerability reports against Express that we are currently working through, which is much higher priority and has our focus at this time. Apologies.

**Comment by dougwilson** [comment]
A quick test seems to have a weird issue, in that the header `accept-encoding: gzip;q=1, br;q=0.3` it sending back br instead of gzip. I would think it should be honoring the quality indicators in the negotiation for the encoding.

**Comment by nicksrandall** [comment]
@danielgindi Thanks for all your work on this. Let me know if you'd like any assistance on addressing the latest issue @dougwilson reported. I'm happy to help where I can.

**Comment by danielgindi** [comment]
> @danielgindi Thanks for all your work on this. Let me know if you'd like any assistance on addressing the latest issue @dougwilson reported. I'm happy to help where I can.

Well if you want to make the explicit quality indicator take precedence, then by all means... :)
I'm kind of busy right now and only reprioritized these because @dougwilson insisted that he does not see any any effect with current browsers.

I'll give you permissions to the repo, so you can complicate things as you wish! ;)

**Comment by Fonger** [comment]
IMO, I think quality indicators support can be released in the upcoming version because most browser does not use it for this moment.

**Comment by dougwilson** [comment]
We of course all know from the above conversation that web browser are not the only clients out there. The PR itself was originally never even going to work for non-browsers, so it seems that the primary use case was for the non-browsers and not the browsers.

**Comment by Fonger** [comment]
Yeah. You're right and I find koa-compress also support quality indicator. We should always follow the standard.

**Comment by dougwilson** [comment]
Yea. I was pointing out the quality stuff just looking at the new commit (it was a change from the previous behavior). I would guess we should define it as "if br is at the highest accepted quality, then prefer that over the others, regardless of ordering". As in `gzip, br` = br and `gzip;q=0.8, br;q=0.8` = br, but `gzip, br;q=0.8` = gzip. It seems like (reading through the old Chrome issues) that Chrome put `br` last knowing servers would not serve it if they followed the standard, as it was "experimental" it... it's still at the end of the list as least preferred even today, and not sure why.

**Comment by nicksrandall** [comment]
I will make a change so that if brotli and gzip are preferred equally (regardless of order) the system will prefer brotli. Otherwise, the system will prefer the more preferred compression algorithm. Sound reasonable?

**Comment by nicksrandall** [comment]
My thinking is to not reinvent the wheel so I've pulled in koa-compress' logic to determine which content encoding header is preferred as it conforms the rules we've specified above. I've also added a test case to verify.

My only concern here is node version support. What is the official node version support for this lib?

**Comment by dougwilson** [comment]
> My only concern here is node version support. What is the official node version support for this lib?

The official Express.js middlewares support the same Node.js versions as the current version of Express.js. The Travis CI has the supported versions listed in the test matrix.

**Comment by dougwilson** [comment]
> I will make a change so that if brotli and gzip are preferred equally (regardless of order) the system will prefer brotli. Otherwise, the system will prefer the more preferred compression algorithm. Sound reasonable?

Yea, that sounds fine to me, as it seems like that is what the web browsers are actually doing for this case.

**Comment by nicksrandall** [comment]
> The official Express.js middlewares support the same Node.js versions as the current version of Express.js.

Forgive me as I'm not super familiar with Travis CI but according to https://expressjs.com/en/changelog/4x.html Express.js seems to only support node 10.x and 12.x but the test matrix goes all the way down to 0.8? Did I interpret something?

**Comment by dougwilson** [comment]
I think you are mistaking the change log for adding support to Node.js versions as supporting _only_ those versions. It supports the same versions the Travis CI matrix of this module supports, so the CI just needs to pass with the existing matrix unless the idea is to hold this PR until the next major version instead of releasing as a semver minor, if that is the desire. Express 5 will be released soon, so it would probably only need to wait a few months if that is what path you want.

**Comment by dougwilson** [comment]
Though it seems like this change has broken the accept in other ways. For example, given the header `identity;q=0` seems to throw an exception instead of sending back the uncompressed response. Also the behavior for the header `gzip;q=0.001` seems to be to no longer compress instead of sending back gzip.

**Comment by nicksrandall** [comment]
Ok, I have all the tests passing on all versions of node, including a new test that tests the "gzip;q=0.001" value.

**Comment by dougwilson** [comment]
Hi @nicksrandall thank you so much! I am still on mobile, and it looks like there is now a substantial amount of code to review 😭 . I'm a bit concerned about why we should be adding such a large amount of code directly to this library, vs adding whatever is needed into Express.js's own `accepts` or other library? It also seems to be in violation of their software license, as it is licensed under MIT, but you did not include the given MIT license and copyright with this substantial part of the software here...

I think (but haven't yet tried) based on the code that it is not as standards compliant as the previous module used here, and we will need to go over this new implementation and compare to the standards with a fine-toothed comb. It seems to me that, for instance, it is not comparing the content coding as a case-insensitive manor as specified in RFC 7231 (but since I'm mobile I haven't tested that -- I am just looking at the code on a small screen 😅 ).

If we want to move forward with this implementation, please give me until the next couple weeks to have the time to circle around to review all of this code... the first few looks I've noticed multiple spec violations, even after you made a bunch of changes, so I'm not sure at what level of detail this large replacement code has been looked over to feel comfortable sending to the 8M+ daily downloads, as we want to really be sure we are not regressing behavior here 🙆‍♂️ .

**Comment by nicksrandall** [comment]
I didn't want to loose what was happening here so I have created an alternate PR (#173) that provides what I believe to be the absolute minimal code change to this library to support brolti in a backwards compatible way.

I'm just hoping that one of these PRs gets merged in the near term. Please let me know which direction you think would be best to continue to pursue.

**Comment by danielgindi** [comment]
@dougwilson I've remove the dependency on koa's modified code, and implemented a simple parsing function for the encoding, with built-in preferred-encoding support. this results in *less* code, and a *faster* code.

Passing all old tests, new tests, and the latest ones added by @nicksrandall 

I really don't see anything blocking a merge now, and it would be great to have this in production :-)

**Comment by dougwilson** [comment]
I'm a bit concerned about why we should be adding such a large amount of code directly to this library, vs adding whatever is needed into Express.js's own `accepts` or other library? We target every library to only have a single `index.js` file if possible, and since we already split out this functionality long ago into these other modules, it seems like putting that functionality there is the correct path forward, rather than putting a large about of code in here that doesn't have a test suite around it.

Besides that, even just putting that code in here seems to have licensing issues, as it seems to be a large chunk of the `negotiator` module, which is licensed under MIT, which requires the original copyright to be kept with substantial portions like the file added here. Just making the changes direct in our dependency (which is part of Express.js) avoids that issues altogether.

**Comment by danielgindi** [comment]
> I'm a bit concerned about why we should be adding such a large amount of code directly to this library, vs adding whatever is needed into Express.js's own `accepts` or other library? We target every library to only have a single `index.js` file if possible, and since we already split out this functionality long ago into these other modules, it seems like putting that functionality there is the correct path forward, rather than putting a large about of code in here that doesn't have a test suite around it.
> 
> Besides that, even just putting that code in here seems to have licensing issues, as it seems to be a large chunk of the `negotiator` module, which is licensed under MIT, which requires the original copyright to be kept with substantial portions like the file added here. Just making the changes direct in our dependency (which is part of Express.js) avoids that issues altogether.

This module uses `accepts` which is not part of the `expressjs` namespace, I do not know who is in charge of that module, but I took a look at that and it seems like it's jumping through hoops for the sake of keeping it generic, and adding a "preferred" functionality will probably not be accepted well there (even though the module is called "accepts"...).  
Note that originally there was a "preferred" functionality implemented here, in `compression`, and not in `accepts`. Although it was only for one "preferred", but still.

I don't see a "large chunk" of `negotiator` module, except for a very small for-loop that extracts a single value from `=` pairs. There are not a lot of ways you could write that piece of code, and I really do not believe that this is an issue in any way. 

Do you want to get `compression` to move forward and support modern compressions or not?

**Comment by dougwilson** [comment]
> This module uses accepts which is not part of the expressjs namespace, I do not know who is in charge of that module

It is part of Express.js: http://expressjs.com/en/resources/community.html#express-is-made-of-many-modules

> and adding a "preferred" functionality will probably not be accepted well there

Why not? I mean, just look at the person who is committing there... it is me. Are you saying I would not accept what I am directing you to put there?

> I don't see a "large chunk" of negotiator module, except for a very small for-loop that extracts a single value from = pairs. There are not a lot of ways you could write that piece of code, and I really do not believe that this is an issue in any way.

Sure, you may have unrolled some of those functions together, but unless you are saying you definately didn't copy and paste from negotiator and just move around the code, then it is certainly the original license at play on that code.

**Comment by danielgindi** [comment]
> Why not? I mean, just look at the person who is committing there... it is me. Are you saying I would not accept what I am suggesting you to put there?

Good to know 😂

> Sure, you may have unrolled some of those functions together, but unless you are saying you definately didn't copy and paste from negotiator and just move around the code, then it is certainly the original license at play on that code.
What happened is actually more simple: I took a look at the regex, did not really trust it as people tend to let regexes run free. So I wrote one carefully on http://regexpal.com/, and grouped it the same way. Then I copied specifically the `param` split part. All the other code was written by me in a few minutes.

I do not really get the insistency on splitting everything to a million pieces, I don't see a value to any extremity, but on the other hand I don't mind making a PR to `negotiator` and `accepts`, as long as there's a willingness to move forward.

**Comment by danielgindi** [comment]
@dougwilson If we're at it - why did you filter out `q=0` in `negotiator`? These are valid quality values, just prioritized the lowest.

**Comment by dougwilson** [comment]
> @dougwilson If we're at it - why did you filter out q=0 in negotiator? These are valid quality values, just prioritized the lowest.

This is because when you combine the module's API and spec, that is the correct functionality. The API of the module only returns what is acceptable. The spec says that 0 is a special value of not acceptable (vs just the lowest acceptability, which would be 0.001). You can find the spec in RFC 7231.

> If the representation's content-coding is one of the content-codings listed in the Accept-Encoding field, then it is acceptable unless it is accompanied by a qvalue of 0.  (As defined in Section 5.3.1, a qvalue of 0 means "not acceptable".)

**Comment by danielgindi** [comment]
@dougwilson you have a PR pending

**Comment by danielgindi** [comment]
I guess everyone are in vacation now

**Comment by nicksrandall** [comment]
@danielgindi just want to say that this is good work. Thanks for pushing for this.

**Comment by paambaati** [comment]
@dougwilson Would you have thoughts on this PR? I wanted to see if the community can do anything else to get this merged, as there’s been multiple unsuccessful efforts to add `brotli` support to this package.

This feature would be massively beneficial to express itself and a host of other huge projects built on top of it (Next.js being one), so we’d love it if there’s some traction on this.

**Comment by manniL** [comment]
@dougwilson Any concerns regarding the current state of the PR?

**Comment by danielgindi** [comment]
At this point I'm convinced this is not going to happen.

**Comment by Kle0s** [comment]
> At this point I'm convinced this is not going to happen.

Would you like to create an npm package from your fork? Or should I? I really think it's much needed (and would like to use it myself)

**Comment by nicksrandall** [comment]
If it is helpful, I’ve built this package and am using it in production in several places: https://github.com/nicksrandall/compression

**Comment by mmahalwy** [comment]
any luck with this? @dougwilson

**Comment by Fonger** [comment]
Chrome has supported brotli since April 14, 2016 (v50) and [koa-compress](https://github.com/koajs/compress) has supported brotli since 28 April, 2020.

I think it's time to move forward for express ecosystem.

Is this library deprecated?

**Comment by maggie0002** [comment]
> If it is helpful, I’ve built this package and am using it in production in several places: https://github.com/nicksrandall/compression

Tempting, thanks for this. How has it been running, any issues?

**Comment by romgrk** [comment]
It is quite infuriating that the core express package for compression hasn't been able to merge brotli support for over 5 years. Please either get this going or mark the package as deprecated and point users to more active packages. There hasn't been a commit in 3 years on this repo.

cc @dougwilson, because you've been the only maintainer here for a while

**Review by vinayakkulkarni — APPROVED** [review]
LGTM

**Comment by nithin-murali-arch** [comment]
Looks like the PR is approved, could you folks resolve the conflicts and merge? Is there anything else we're waiting on?

**Comment by dgautsch** [comment]
Any plans on having this merged or is https://www.npmjs.com/package/shrink-ray-current the current best solution?

**Comment by barnhill** [comment]
Is there any current movement on getting this approved and merged?  Im asking in reference to Apollo Server which we use for GraphQL.  It doesn't support Brotli due to this dependency not supporting it.

**Comment by MannyPamintuanAtHeb** [comment]
@dougwilson, I was curious if you have the bandwidth to review this PR and provide a secondary approval with @vinayakkulkarni having already reviewed and approved?

Seeing `br` support land for Express would be impactful to `apollo-server-express` which currently at this time does not support `br` for compression.

Anything that the community can do to help see this PR land and become a reality?

**Comment by romgrk** [comment]
Realistically this repository should be considered abandoned, the maintainers haven't responded in a long time and brotli has been out for around 8 years so it's safe to assume there is no interest to add it. The community should probably rally around one of the multiple existing forks for this package. Maybe [express-compression](https://github.com/gumlet/express-compression) would be good?

**Comment by bjohansebas** [comment]
@danielgindi @nicksrandall Thank you for this work, it has been used as a reference to move forward with #194. You can continue helping by reviewing #194 so we can launch it soon.

**Pull request closed without merging**
